### PR TITLE
refactor: consolidated ABI stores, project config hierarchy, directory CI runs

### PR DESCRIPTION
**Pull request by tamtamchik**
## What

Restructures every config into a per-deployment hierarchy, consolidates ABI storage into one file per directory, and hardens the tool around it: chain verification on both RPC and explorer sides, on-chain implementation checks for every proxy entry, and CI that runs whole directories.

## Config layout

- `configs/<project>/<mainnet|testnet>/` per deployment: `lido`, `lido-multichain`, `lido-earn`, `mellow`, `defiwrapper`, `hypernative-guard`. Filenames carry product and network when a directory spans several.
- Anchor conventions, enforced across all 27 configs: addresses are `lowerCamelCase`, constants are `UPPER_SNAKE`, every `*_ROLE` hash lives in the `roles:` section (`DEFAULT_ADMIN_ROLE` first, grouped per contract), contracts that need ABIs keep their addresses in `deployed:`.
- Config content: mainnet committees are back under checks (`circuitBreakerCommittee`, `csmCommittee`, `stVaultsCommittee`), `VaultsAdapter` and the TW exit-hash factories are verified, Safes are modeled as proxies pinned to their singletons, `ProxyAdmin` entries are replaced by the `proxyAdminOwner:` field, transient exit-limit tuples are unpinned.

## Multichain scope

`lido-multichain` now covers only the networks documented under [Lido Multichain](https://docs.lido.fi/deployed-contracts#lido-multichain) — Arbitrum, Optimism, Base, Linea, BSC, Unichain. The eleven mainnet configs for networks listed elsewhere (ink, lisk, mantle, megaeth, mode, monad, plasma, soneium, swell, zircuit, zksync) and the soneium testnet config are removed, their ABI store entries pruned, and `.env.sample` trimmed to the RPC endpoints configs actually name.

## Pruned checks

Beyond the removed networks, the migration drops checks under one policy: a config asserts what its deployment holds today, so vote snapshots, retired scopes and pins subsumed by stronger checks do not migrate.

- The `sr2`, `tw` and `optimism` before/after-voting snapshot configs are gone: the votes are enacted, and the resulting state is what the `lido` and `lido-multichain` configs assert.
- `dvv` re-lands as `configs/mellow/mainnet/dvv.yaml`, scoped to the vault and its admins (~96 paths fewer).
- `mellow_strategy` merges into the `lido-earn` strategy configs (~58 paths fewer): one factory implementation, `0x0000000397b71C8f3182Fd40D247330D218fdC72`, sits behind every instance and is pinned per factory via `implementationAt`, so the per-entity duplicates are dropped.
- The `wsteth-optimism` bridges lose their `l1Token`/`l2Token`/`messenger` view pins: the values are constructor immutables of the bridge implementations, which the config now pins on-chain.
- The `GateSealFactory` entry (one `get_blueprint` check) leaves the mainnet lido config; the deployed GateSeals themselves stay asserted.

## ABI stores

- One `abis.json.gz` per config directory, keyed by `chainId:address` (lowercase), value `{ name, abi }`. 730 entries across 8 stores.
- Every run downloads the ABIs missing from the store — a config that gained an address needs no flag.
- `--update-abi` rebuilds: re-downloads every address the run walks and prunes unreferenced keys once a directory run finishes. An address the explorer refuses to serve (unverified contract, dead explorer) keeps its stored ABI, so a rebuild can never lose one. A store the loader cannot read (any pre-consolidation format) rebuilds from scratch under the flag.
- Single-file runs never prune the shared store; a shared address downloads once per run.

## New checks

- The RPC's `eth_chainId` must match the config `chainId` — a wrong endpoint stops the run instead of green-lighting the wrong network. Explorer hostnames are verified too (skipped for etherscan v2, where the chain is a request parameter).
- `implementation:` is verified on-chain for every entry (EIP-1967 slot → `implementation()` → Safe singleton slot, the last gated to `SafeProxy`/`GnosisSafeProxy`). An EIP-1967 proxy described as a regular contract fails; an undeclared Safe or getter-only proxy still passes as "not a proxy", since the implementation slot is the only read that can judge an entry the config does not declare a proxy. `--skip-implementation-check` bypasses the layer, and every bypassed entry lands in the skip counter the summary line reports even under `--quiet`. An inconclusive read is a skip with a warning, never a pass — and warnings print under `--quiet`.
- `proxyAdminOwner:` reads the admin slot and calls `owner()` on whatever it finds; a renounced admin (`owner() == 0x0`) is expressible.

## State sync

Configs are realigned with what the chains hold today:

- Hoodi curated module: `STAKING_ROUTER_ROLE` gained a second holder.
- Hoodi: the DepositSecurityModule moves to `0x8E63F0aF403ffd3Cbd5dB18b4ee632314ab49B51`, and the CSM v2 family joins the config — module `0xbb7dd81FAC80f3Effa10eA8b973c15AE65a4CAf9`, accounting `0x04A0294bF3306532309D7DD776D4A7eF502313e0`, fee oracle `0x9B8bBA11bbE1a351CC8dD1CFCa6719FF7274A208`, verifier `0xFdE0FD9aDa4E898D3b34Dd4EA3433b75f0B6dd30`, ejector `0xf8a71C08DBe7D2efaD76D3951a3065B8cE20e4f0`.
- `hypernative-guard` tracks the current activation state per Safe instead of the target state: three of four guards expect `isPassThroughMode: true` (the deploy default) until enforcement is switched on for them.
- Mainnet earn vaults: Mellow added redeem and sync-redeem queues to earnUSD, earnUSDc, earnUSDe and the strETH strategy vault, moving queue counts, limits and hooks; verifier merkle roots are refreshed and `SET_MERKLE_ROOT_ROLE` lost a holder on the earnETH vault.
- `steth-optimism`: LidoLocator points at its current implementation, and `postTokenRebaseReceiver` at the new `TokenRateNotifier`, whose `observers` now return `[addr, kind]` and whose interface constant split in two.

## Reliability

- Explorer requests are paced to 3 req/s process-wide and fired concurrently; HTTP 429 and JSON-style rate limits get one retry; implementation chains that loop are cut.
- Transient RPC failures (408/429/5xx/timeout) retry once at the provider level — public-endpoint flakes stop failing CI.
- CI runs each config directory as a matrix job on Node 24; secrets append to `.env` so a missing one falls back to the `.env.sample` default.
- Endpoints that stopped serving us are replaced: the Sepolia default moves to publicnode after drpc made the chain paid-only, and `steth-optimism` reads its ABIs from etherscan after `optimism.blockscout.com` stopped confirming its chain id.

## Removed

- `--generate` and seed configs (unsound on explorer failures, unused).
- Sourcify support and the per-explorer handler files — one client covers etherscan v2, blockscout and etherscan-compatible hosts.
- The stale `.skills/` duplicate of the state-mate skill.
- `chainId` now rejects anything but a positive integer.

### CODE REVIEW

File: configs/lido-earn/mainnet/strategy-arbitrum.yaml
Comment (TheDZhon) [inline configs/lido-earn/mainnet/strategy-arbitrum.yaml:1]: not sure about this file changes though

File: configs/lido-earn/mainnet/strategy-plasma.yaml
Comment (TheDZhon) [inline configs/lido-earn/mainnet/strategy-plasma.yaml:1]: also quite not clear why single impl everywhere

File: configs/lido-earn/mainnet/strategy-ethereum.yaml
Comment (TheDZhon) [inline configs/lido-earn/mainnet/strategy-ethereum.yaml:1]: 🤔

File: src/section-validators/base.ts [hunk 1]
```diff
@@ -101,6 +104,15 @@ export abstract class SectionValidatorBase {
     basePath?: string,
   ): Promise<void>;
 
+  /**
+   * For proxy contracts, the checks run against the implementation ABI
```
Comment (TheDZhon) [inline src/section-validators/base.ts:95]: might be tricky for TPU btw, as TPUs have additional abi that is available only for ones who are the proxy admin

File: README.md [hunk 2]
```diff
@@ -128,44 +128,23 @@ l1:
...
-- Places the output file alongside your config file
-- Provides compression statistics
-
-**Note**: You cannot use both consolidated and individual ABI files simultaneously. state-mate will automatically detect which format you're using.
+The chain ID distinguishes contracts deployed at the same address on different networks. The `name:` / `proxyName:` fields in YAML serve as a sanity check: state-mate compares them with the stored name and fails on mismatch. For proxies, `checks` resolve the ABI at `implementation:`, `proxyChecks` at `address:`. See [configs](/configs/).
 
 #### Updating ABIs
 
-Use the `--update-abi` option to download all ABIs, overwriting existing files:
+`--update-abi` downloads the ABIs missing from `abis.json.gz` and leaves existing entries alone:
```
Comment (TheDZhon) [inline README.md]: why alone?

**Review by TheDZhon — COMMENTED** [review]
Good and juicy 👏 

Raised a few questions and considerations to proceed on this 👍

**Comment by TheDZhon** [comment]
I guess we should pin `chainId` in every cfg from now on (and probably in the old ones too) for better compat with forked networks

File: src/state-mate.ts
Comment (TheDZhon) [inline src/state-mate.ts:1]: I think here we have a slight regression (it was already a bug but materialized more with this PR)

`--update-abi` can fail to populate a missing implementation ABI when the proxy ABI already exists. 

Here in the file the updater skips the explorer call as soon as `chainId:proxyAddress` exists, so it never discovers `Implementation`. 

So minimal fix is to remove the pre-explorer skip in `iterateLoadedContracts` and let `checkAllAbi` skip existing keys while still saving missing implementations; also worth adding an integration unit test for "existing proxy ABI, missing implementation ABI".

Comment (TheDZhon) [reply src/state-mate.ts:1]:
```
diff --git a/src/state-mate.ts b/src/state-mate.ts
index 247cfb0..1afef79 100644
--- a/src/state-mate.ts
+++ b/src/state-mate.ts
@@ -10,7 +10,7 @@ import chalk from "chalk";
 import { JsonRpcProvider } from "ethers";
 import * as YAML from "yaml";
 
-import { checkAllAbi, flushAbiUpdates, getAbiNameForAddress } from "./abi-provider";
+import { checkAllAbi, flushAbiUpdates } from "./abi-provider";
 import { doGenerateBoilerplate } from "./boilerplate-generator";
 import { parseCommandLineArguments } from "./cli-parser";
 import { printError, readUrlOrFromEnvironment, resolveChainId } from "./common";
@@ -184,11 +184,6 @@ async function iterateLoadedContracts<T extends EntireDocument | SeedDocument>(
         log(`\n${WARNING_MARK} ${chalk.yellow(`The env var ${explorerTokenEnv} is not set`)}\n`);
       }
       for (const address of addresses) {
-        const existingAbiName = getAbiNameForAddress(chainId, address);
-        if (existingAbiName !== undefined) {
-          log(`ABI ${chalk.magenta(`${existingAbiName} @ ${address}`)} ${chalk.green("Skipped (exists)")}`);
-          continue;
-        }
         const contractInfo = await loadContractInfoFromExplorer(address, explorerHostname, explorerKey, chainId);
         if (!contractInfo) {
           continue;
diff --git a/tests/abi-provider.test.ts b/tests/abi-provider.test.ts
index 6f3ff43..4a40fa0 100644
--- a/tests/abi-provider.test.ts
+++ b/tests/abi-provider.test.ts
@@ -148,12 +148,29 @@ describe("loadAbiFromFile", () => {
 });
 
 describe("downloadAndCheckAllAbi", () => {
-  it("does not query the explorer when the ABI already exists", async () => {
-    setupConfigDirectory({
+  it("keeps an existing proxy ABI while saving a missing implementation ABI", async () => {
+    const directory = setupConfigDirectory({
       [scopedKey(ETH_CHAIN_ID, PROXY_ADDRESS)]: { name: "OssifiableProxy", abi: PROXY_ABI },
     });
-    const fetchMock = mock.method(globalThis, "fetch", async () => {
-      throw new Error("unexpected explorer request");
+    const fetchMock = mock.method(globalThis, "fetch", async (url: Parameters<typeof fetch>[0]) => {
+      const requestUrl = String(url).toLowerCase();
+      const responseAbi = requestUrl.includes(IMPL_ADDRESS.toLowerCase()) ? LIDO_ABI : PROXY_ABI;
+      const responseName = requestUrl.includes(IMPL_ADDRESS.toLowerCase()) ? "Lido" : "OssifiableProxy";
+      return {
+        ok: true,
+        json: async () => ({
+          status: "1",
+          message: "OK",
+          result: [
+            {
+              ABI: JSON.stringify(responseAbi),
+              ContractName: responseName,
+              IsProxy: responseName === "OssifiableProxy" ? "true" : "false",
+              Implementation: responseName === "OssifiableProxy" ? IMPL_ADDRESS : "",
+            },
+          ],
+        }),
+      } as Response;
     });
 
     try {
@@ -161,7 +178,11 @@ describe("downloadAndCheckAllAbi", () => {
         deployed: { l1: [PROXY_ADDRESS] },
         l1: { rpcUrl: "http://localhost:1", explorerHostname: "api.etherscan.io", chainId: 1 },
       } as SeedDocument);
-      assert.equal(fetchMock.mock.callCount(), 0);
+      assert.equal(fetchMock.mock.callCount(), 2);
+
+      const store = readStore(directory);
+      assert.deepEqual(store[scopedKey(ETH_CHAIN_ID, PROXY_ADDRESS)], { name: "OssifiableProxy", abi: PROXY_ABI });
+      assert.deepEqual(store[scopedKey(ETH_CHAIN_ID, IMPL_ADDRESS)], { name: "Lido", abi: LIDO_ABI });

     } finally {
       fetchMock.mock.restore();
     }
@@ -189,7 +210,7 @@ describe("downloadAndCheckAllAbi", () => {
         l2: { rpcUrl: "http://localhost:2", explorerHostname: "api.bscscan.com", chainId: BSC_CHAIN_ID },
       } as SeedDocument);
 
-      assert.equal(fetchMock.mock.callCount(), 1);
+      assert.equal(fetchMock.mock.callCount(), 2);
       const store = readStore(directory);
       assert.deepEqual(store[scopedKey(ETH_CHAIN_ID, PROXY_ADDRESS)], {
         name: "ERC1967Proxy",
```

**Review by TheDZhon — COMMENTED** [review]
🤔

**Comment by tamtamchik** [comment]
Chain Id now required.

File: README.md (hunk 2 above)
Comment (tamtamchik) [reply README.md]: So by default update-abi, and --update-abi will be drop all and re-download

File: configs/lido-earn/mainnet/strategy-plasma.yaml
Comment (tamtamchik) [reply configs/lido-earn/mainnet/strategy-plasma.yaml:1]: In fact it was one Factory implementation, `0x0000000397b71C8f3182Fd40D247330D218fdC72`, behind all instances. The old configs named the entity implementation there and got away with it because ABI lookup went by contract name and it goes by address now. Entity implementations now stayed where they belong, in each factory's `implementationAt`. And `implementation:` is verified against the chain for every entry now.

Comment (tamtamchik) [reply configs/lido-earn/mainnet/strategy-plasma.yaml:1]: Fixed in b0e9467

File: configs/lido-earn/mainnet/strategy-ethereum.yaml
Comment (tamtamchik) [reply configs/lido-earn/mainnet/strategy-ethereum.yaml:1]: Fixed in b0e9467

File: src/section-validators/base.ts (hunk 1 above)
Comment (tamtamchik) [reply src/section-validators/base.ts:95]: Right, and there is nothing to call anyway. Hence `proxyChecks: {}` on every TPU in the repo.

The implementation comes from the EIP-1967 slot for every entry, and the new optional `proxyAdminOwner`: reads the admin slot and calls `owner()` on whatever sits there, so the ProxyAdmin needs no anchor and no ABI. Replaced 124 standalone ProxyAdmin entries with new guard.

**Review by TheDZhon — COMMENTED** [review]
codex:

1. **ABI provenance / CI integrity**

   CI now trusts every existing archive entry and only downloads missing entries. Because method coverage is derived from that archive, a stale or edited ABI can silently remove or change the methods CI believes it checked; the gzip diff is also opaque to review. Please make the committed store a verified input: deterministically refetch/compare every referenced ABI (or verify a canonical text/hash manifest), fail when a normal/CI run mutates a store, and expose the semantic ABI delta.

2. **Skipped checks hidden by quiet mode**

   Every directory job now uses `--quiet`, but `_checkViewFunction` increments the total before a `result: null` returns, and `logMethodSkipped` emits nothing in quiet mode. That makes skipped methods appear in “checks passed” while hiding the skip list. Please track checked/skipped separately, keep the skip count visible in quiet CI, and require an explicit policy/allowlist for skips; a whole-result `null` should not count as passed.

3. **RPC failure misdiagnosed as missing admin**

   `readSlotWord` returns `{ failed: true }` when an RPC storage read throws, but `checkProxyAdminOwner` ignores that flag and reports that the proxy keeps no EIP-1967 admin. I reproduced this with a provider that rejected/rate-limited batch calls; the same proxy passed through another provider and a direct slot read returned its configured admin. Please report an indeterminate storage-read failure/retry when `failed` is set, and add a test distinguishing a thrown read from a successfully read empty slot.

4. **Retired monitoring surfaces**

   The official deployment inventory still records governance/emergency surfaces for several removed networks, even if they are absent from the narrower current multichain subsection. What durable monitor or owner replaces each deleted config? Please add a retirement manifest/tombstone linked to the replacement, or retain the config until that replacement exists.

5. **Reproducible state delta**

   This migration also changes holders, roots, implementations, queues, and config membership, while CI verifies floating current state. Please publish a machine-readable non-mechanical delta with chain ID, pinned block/hash, and exact reproducing commands so reviewers can distinguish the intended state sync from path/archive churn.

**Comment by tamtamchik** [comment]
@TheDZhon Thanks! 2 and 3 were real defects, both fixed in 750a610. Other issues wontfix :)

**Comment by TheDZhon** [comment]
New codex findings including stacked changes:

1. PR #156 is not merge-ready. All five added configurations failed against two independent RPC-provider sets. A pinned Optimism check confirmed that the configuration monitors the retired automation pair while omitting the active pair recorded by the source repository. Compare State Mate’s configuration with the current deployment record.
**Required action**: regenerate all five configurations from a pinned source revision and pinned blocks, include both active and deliberately retained retired contracts, and require the new CI job to pass.
2. PR #171’s implementation looks sound, but the stacked documentation still omits the new `proxyAdmin` field and recommends the weaker owner-plus-storage approach. Update the example and field table in `configuration.md` and the guidance in `abi-and-proxies.md`.
3. PR #152 incorporated two earlier findings: skipped checks are counted separately, and unreadable admin-slot calls no longer look like empty slots. The ABI provenance problem remains: ordinary runs trust an existing archive without comparing it to a canonical source, while CI runs only the ordinary path. See `state-mate.ts`, `abi-provider.ts`, and `ci.yaml`.

**Comment by TheDZhon** [comment]
Claude:

**#152 should change in four points before merge:**

1. Declare the full delta in the body. The migration also removes 183 check paths outside the declared list: the sr2, tw, and optimism before/after-voting snapshot configs; ~96 dvv and ~58 mellow-strategy paths; the Optimism bridge `l1Token`/`l2Token`/`messenger` pins; one GateSealFactory check. Two expected-value changes are also undeclared: HypernativeGuard `isPassThroughMode` false→true (confirmed true on-chain at block `25838992` on two providers) and the new Hoodi DSM/CS02 sync. All look intentional. None is declared. State the pruning policy so intent is separable from accident.
2. Make `--skip-implementation-check` visible. It bypasses the whole implementation layer with no trace (`src/section-validators/implementation.ts:91`), and the new `skipped` counter does not cover it. Print one quiet-proof warning line or route it through `incSkipped`.
3. Align the body with the code: "a proxy described as a regular contract fails" holds only for EIP-1967 proxies. An undeclared Safe or getter-only proxy passes as "not a proxy" (`implementation.ts:177-181`, deliberate). #169's `docs/abi-and-proxies.md` already states the boundary correctly.
4. On ABI provenance (per the codex comment): the current stores audit clean, so the gap is prospective, not present. Run the #169 daily cron with `--update-abi`, or add a compare-only `--verify-abi` mode. A `--update-abi` directory pass before merge also drops the 5 stale orphan keys in the lido-earn store.

Pre-existing, for a follow-up issue: `mustRevert` counts any throw as a revert (`base.ts:162-169`), so a dead RPC turns must-revert checks into passes.

**Stack.** 
- #171: advance. The checker is fail-closed with no bypass flag; 6 of 6 sampled pins match the admin slot on two providers; 3 of 3 `cast compute-address <proxy> --nonce 1` derivations reproduce the pin. Update `configuration.md` and `abi-and-proxies.md`, which still recommend the replaced `storage:` approach. 
- #169: fine; note `engines.node >=24` breaks Node 22 environments. 
- #156: hold, confirming the codex finding with pins. The asserted canary state has already migrated on-chain: L1 owner and ADMIN are the DAO Agent, the sender points at the new pool, `getDelay` is `43200`, and the old automation's `SYNC_ROLE` is revoked while no sampled address holds `SYNC_ROLE`. One item beyond the codex findings: the new OraclePool, SyncTrigger, and CREReceiver are owned by `0xFc832dA3D688352C0aB1A32136c7fABbB16d66E6`. This address appears nowhere in the config, the `migration:`/`handoff:` comments, or the documented handoff targets. Identify this owner before the configs are regenerated.

**Comment by tamtamchik** [comment]
@TheDZhon for this PR:
- Bypassed entries land in the skip counter in 109a65b.
- The PR description updated.
- Orphan keys pruned in 4a9ca7d.
- mustRevert filed as #173.

I think we can merge this one

**Review by TheDZhon — APPROVED** [review]
